### PR TITLE
add TDM-GCC64 build

### PR DESCRIPTION
**Pull request by yami-hack**
I try to add the TDM-GCC64 build, the speed of running under the window to improve a lot, but my CPU does not support SSE4.1, so the modified into SSE3
My English is poor

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -69,8 +69,11 @@ if (NOT MSVC)
     set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS}")
 
     if (ARCHITECTURE_x86_64)
-        set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -msse4.1")
-        set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -msse4.1")
+        #set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -msse4.1")
+        #set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -msse4.1")
+        #Some CPU do not support sse4.1
+        set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -msse3")
+        set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -msse3")
```
Comment (JayFoxRox) [inline CMakeLists.txt]: Make this an option at least

File: CMakeLists.txt [hunk 2]
```diff
@@ -114,11 +117,16 @@ add_definitions(-DSINGLETHREADED)
...
+if(ENABLE_SUPPORT_PNG)
+    find_package(PNG QUIET)
+    if (PNG_FOUND)
+        add_definitions(-DHAVE_PNG)
+    else()
+        message(STATUS "libpng not found. Some debugging features have been disabled.")
+    endif()
 else()
-    message(STATUS "libpng not found. Some debugging features have been disabled.")
+
```
Comment (JayFoxRox) [inline CMakeLists.txt]: Why is this an empty scope now?

File: CMakeLists.txt [hunk 3]
```diff
@@ -114,11 +117,16 @@ add_definitions(-DSINGLETHREADED)
...
 endif()
```
Comment (JayFoxRox) [inline CMakeLists.txt]: Why is this an empty scope now?

File: CMakeLists.txt [hunk 4]
```diff
@@ -172,7 +194,9 @@ IF (APPLE)
 ELSEIF(MINGW)
     # GCC does not support codecvt, so use iconv instead
     # PSAPI is the Process Status API
-    set(PLATFORM_LIBRARIES winmm ws2_32 psapi iconv)
+    # TDM-GCC64 not support iconv
```
Comment (JayFoxRox) [inline CMakeLists.txt]: `# TDM-GCC64 does not support iconv`

File: CMakeLists.txt [hunk 5]
```diff
@@ -172,7 +194,9 @@ IF (APPLE)
...
+    # set(PLATFORM_LIBRARIES winmm ws2_32 psapi iconv)
```
Comment (JayFoxRox) [inline CMakeLists.txt]: dead code

File: TDMGCC64_SDL2_build.md [hunk 6]
```diff
@@ -0,0 +1,13 @@
+
+TDM-GCC64 build 
+
+
+bash
+git clone --recursive https://github.com/yami-hack/citra.git
```
Comment (JayFoxRox) [inline TDMGCC64_SDL2_build.md]: Don't add other peoples repo to the master

File: externals/microprofile/microprofile.h [hunk 7]
```diff
@@ -923,15 +923,18 @@ void MicroProfileThreadJoin(MicroProfileThread* pThread)
...
-    return (uint32_t)F(0);
+    //return (uint32_t)F(0);
+	void *ret = F(0);
+    return *(uint32_t*)&ret;
 }
 
 void MicroProfileThreadStart(MicroProfileThread* pThread, MicroProfileThreadFunc Func)
 {
-    *pThread = CreateThread(0, 0, ThreadTrampoline, Func, 0, 0);
+    //*pThread = CreateThread(0, 0, ThreadTrampoline, Func, 0, 0);
```
Comment (JayFoxRox) [inline externals/microprofile/microprofile.h]: More dead code

File: externals/microprofile/microprofile.h [hunk 8]
```diff
@@ -923,15 +923,18 @@ void MicroProfileThreadJoin(MicroProfileThread* pThread)
...
+	*pThread = CreateThread(0, 0, ThreadTrampoline, (void*)Func, 0, 0);
```
Comment (JayFoxRox) [inline externals/microprofile/microprofile.h]: C-Style casts are not allowed

File: TDMGCC64_SDL2_build.md [hunk 9]
```diff
@@ -1,13 +1,15 @@
...
 
-
+### Building Citra (Optimized):
+```
 bash
 git clone --recursive https://github.com/yami-hack/citra.git
 mkdir build
 cd build
-cmake ../citra -DCMAKE_BUILD_TYPE=Release -G "MinGW Makefiles" \
+#cmake ../citra -DCMAKE_BUILD_TYPE=Release -G "MinGW Makefiles" \
```
Comment (JayFoxRox) [inline TDMGCC64_SDL2_build.md]: What's the `#` good for?

**Comment by MerryMage** [comment]
Ping @darkf due to mingw-related changes.

File: src/common/string_util.cpp [hunk 10]
```diff
@@ -363,6 +377,41 @@ std::string CP1252ToUTF8(const std::string& input)
 
 #else
 
+#if defined(__MINGW32__)
+// MINGW32 and TDM-GCC64
+static std::wstring CPToUTF16(u32 code_page, const std::string& input)
+{
+    auto const size = MultiByteToWideChar(code_page, 0, input.data(), static_cast<int>(input.size()), nullptr, 0);
+
+    std::wstring output;
```
Comment (lioncash) [inline src/common/string_util.cpp]: std::wstring output(size, '\0');

File: src/common/string_util.cpp [hunk 11]
```diff
@@ -363,6 +377,41 @@ std::string CP1252ToUTF8(const std::string& input)
...
+
+std::wstring UTF8ToUTF16W(const std::string &input)
+{
+    return CPToUTF16(CP_UTF8, input);
+}
+
+std::string UTF16ToUTF8(const std::wstring& input){
+    auto const size = WideCharToMultiByte(CP_UTF8, 0, input.data(), static_cast<int>(input.size()), nullptr, 0, nullptr, nullptr);
+
+    std::string output;
```
Comment (lioncash) [inline src/common/string_util.cpp]:
```cpp
std::string output(size, '\0');
```

**Comment by JayFoxRox** [comment]
This had potential I think? No reason to close it imho.

File: src/common/string_util.cpp (hunk 11 above)
Comment (yami-hack) [inline src/common/string_util.cpp]: I have already created the branch of the project,buildp is mainly to build a platform,and test

**Pull request closed without merging**